crates/unftp-auth-jsonfile/src/lib.rs

```diff
@@ -71,25 +91,37 @@ impl JsonFileAuthenticator {
     /// Initialize a new [`JsonFileAuthenticator`] from json string.
     pub fn from_json<T: Into<String>>(json: T) -> Result<Self, Box<dyn std::error::Error>> {
         let db: Vec<Credentials> = serde_json::from_str::<Vec<Credentials>>(&json.into())?;
-        let salts: BTreeSet<String> = db.iter().map(|credential| credential.pbkdf2_salt.clone()).collect();
-        if db.len() != salts.len() {
-            return Err(Box::new(AuthenticationError::new("The provided salts for the JsonFileAuthenticator must be unique.")));
-        }
         Ok(JsonFileAuthenticator {
             db: db
                 .into_iter()
-                .map(|user_info| {
-                    (
-                        user_info.username,
-                        Password {
-                            pbkdf2_salt: user_info.pbkdf2_salt,
-                            pbkdf2_key: base64::decode(user_info.pbkdf2_key)
-                                .expect("Could not base64 decode the key")
+                .map(|user_info| match user_info {
+                    Credentials::Plaintext { username, password } => (username, Password::PlainPassword { password }),
+                    Credentials::Pbkdf2 {
+                        username,
+                        pbkdf2_salt,
+                        pbkdf2_key,
+                        pbkdf2_iter,
+                    } => (
+                        username.clone(),
+                        Password::Pbkdf2Password {
+                            pbkdf2_salt: base64::decode(pbkdf2_salt)
+                                .expect("Could not base64 decode the salt")
```

Comment:
The containing function already returns `Box<dyn std::error::Error>` so using operator `?` should be easy except the declarations withing the closure blocks us here, might want to use an imperative style (i.e. not use `map`) or have the closure return a `Result<T, Box<dyn std::error::Error>>` too

Reply:
I tried this:

```
    /// Initialize a new [`JsonFileAuthenticator`] from json string.
    pub fn from_json<T: Into<String>>(json: T) -> Result<Self, Box<dyn std::error::Error>> {
        let db: Vec<Credentials> = serde_json::from_str::<Vec<Credentials>>(&json.into())?;
        let map: Result<HashMap<String, Password>, _> = db
            .into_iter()
            .map(|user_info| JsonFileAuthenticator::creds_to_username_pw_pair(user_info))
            .collect();
        Ok(JsonFileAuthenticator { db: map? })
    }

    fn creds_to_username_pw_pair(user_info: Credentials) -> Result<(String, Password), Box<dyn std::error::Error>> {
        let r = match user_info {
            Credentials::Plaintext { username, password } => (username, Password::PlainPassword { password }),
            Credentials::Pbkdf2 {
                username,
                pbkdf2_salt,
                pbkdf2_key,
                pbkdf2_iter,
            } => (
                username.clone(),
                Password::Pbkdf2Password {
                    pbkdf2_salt: base64::decode(pbkdf2_salt)?.try_into()?,
                    pbkdf2_key: base64::decode(pbkdf2_key)?
                        .validate("pbkdf2_key", &Length::Max(SHA256_OUTPUT_LEN))
                        .result()?
                        .unwrap()
                        .try_into()?,
                    pbkdf2_iter,
                },
            ),
        };
        Ok(r)
    }

```